feat(signin): Signin confirmation feature only

vbudhram opened the pull request:
This PR contains only the signin confirmation feature. It is the second PR of https://github.com/mozilla/fxa-auth-server/pull/1232.

shane-tomlinson commented:
@vbudhram - can you rebase?

shane-tomlinson commented on .env.dev:
```diff
@@ -9,5 +9,5 @@ CUSTOMS_SERVER_URL=none
 LOCKOUT_ENABLED=true
 SNS_TOPIC_ARN=disabled
-TRUSTED_JKUS=http://127.0.0.1:8080/.well-known/public-keys,http://127.0.0.1:10139/.well-known/public-keys
+TRUSTED_JKUS=http://127.0.0.1:8080/.well-known/public-keys,http://127.0.0.1:10139/.well-known/public-keys,http://127.0.0.1:9000/.well-known/public-keys
```
Is the new trusted jku supposed to be added?

vbudhram replied:
This was brought over from original PR, but I think it can be removed. It was added so that the `mail-dev` tool could intercept emails for visual inspection.

vbudhram commented on lib/routes/account.js:
```diff
@@ -946,10 +1013,33 @@ module.exports = function (
+          var doSigninConfirmation = requestHelper.shouldEnableSigninConfirmation(sessionToken.uid.toString('hex'), config)
+
+          if (doSigninConfirmation) {
+            return db.sessionTokenWithVerificationStatus(sessionToken.id)
```
@rfk You mentioned looking up the verification status from the auth strategy, https://github.com/mozilla/fxa-auth-server/pull/1232#discussion_r63462831, however, IIUC that strategy will reject the request as [unauthorized](https://github.com/mozilla/fxa-auth-server/blob/48a6fa6335a29dab371d68c31507408ebd655a0d/lib/server.js#L69). Is there another way to get the verification status from auth strategy?

cc @philbooth

philbooth replied:
Oh, good point @vbudhram. Maybe there is a case for two separate strategies, `sessionTokenWithVerificationStatus` and `sessionTokenWithVerificationStatusStrict` (or something, that name is off the top of my head)? So the basic strategy would be created via `makeCredentialsFn` and the strict one would be created via `makeVerifiedCredentialsFn`.

Reasonable suggestion or is there a simpler solution?

rfk replied:
The comment on `sessionTokenWithVerificationStatus` says:

```
// This strategy fetches the sessionToken with its
// verification state. It doesn't check that state.
```

But it looks like that's misleading, and the code actually *does* check the state.

rfk replied:
So stepping back, what route is currently using the `sessionTokenWithVerificationStatus` strategy and depending on it rejecting unverified tokens?  If we need both, I suggest something like:

* `sessionTokenWithVerificationStatus` -> looks up the status, but doesn't assert it's any particular value
* `sessionTokenVerified` -> we had this name originally and I think it's a good name for when we are in fact insisting that the the session is verified

But it's not clear to me what routes would use a strict `sessionTokenVerified` strategy.

philbooth replied:
> But it's not clear to me what routes would use a strict `sessionTokenVerified` strategy.

Oh, excellent point. I think the use of `makeVerifiedCredentialsFn` is just hangover from before we shifted the responsibility for rejection into the token server. It should be fine to just fix `sessionTokenWithVerificationStatus` I think.

vbudhram replied:
Looks like no routes use the `sessionTokenWithVerificationStatus` strategy. I think it is fair to update `sessionTokenWithVerificationStatus` to correctly return the token status and match the comment description.

Will update!

shane-tomlinson commented on lib/routes/account.js:
```diff
@@ -1013,41 +1118,77 @@ module.exports = function (
+                  })
+                })
+                .catch(function (err) {
+                  if (err.errno === error.ERRNO.INVALID_TOKEN && butil.buffersAreEqual(code, account.emailCode)) {
```
There is an unhandled error somewhere here. If the user creates an account with the old auth server and verifies with the new auth server, the `invalidVerificationCode` response is returned to the user and the "Verification Link Damaged" screen is displayed.

The errno being returned is `116` (GONE?), INVALID_TOKEN is `110`

The auth-db-server returns `116` for `notFound`, both db's use `notFound` if the token is not found:

* https://github.com/mozilla/fxa-auth-db-mysql/blob/master/lib/db/mem.js#L322
* https://github.com/mozilla/fxa-auth-db-mysql/blob/master/lib/db/mysql.js#L537
* https://github.com/mozilla/fxa-auth-db-mysql/blob/master/fxa-auth-db-server/lib/error.js#L36

shane-tomlinson replied:
I think the cause of the confusion is that the auth-server and auth-db-server use the same errno to mean two different things. errno 116 on the auth-db-server is `notFound`, errno 116 on the auth-server is [ENDPOINT_NOT_SUPPORTED](https://github.com/mozilla/fxa-auth-server/blob/37b6f49c07260a36a40c7b9eb20c9d197c3a7baf/lib/error.js#L17). errno 110 [does not exist in the auth-db-server](https://github.com/mozilla/fxa-auth-db-mysql/blob/master/fxa-auth-db-server/lib/error.js), but means [INVALID_TOKEN](https://github.com/mozilla/fxa-auth-server/blob/37b6f49c07260a36a40c7b9eb20c9d197c3a7baf/lib/error.js#L26) on the auth server.

shane-tomlinson replied:
FWIW, the signup w/ old auth-server->verify with new-auth server tests pass if I change this line to:

```js
 if (err.errno === 116) {
```

Oddly the `buffersAreEqual` test keeps failing, even though when I print both codes they are identical 
```js
butil.buffersAreEqual(code, account.emailCode)
```

I display both codes using:
```js
                  console.log('code', code.toString('hex'));
                  console.log('emailCode', account.emailCode.toString('hex'));
```

vbudhram replied:
I don't believe we can remove the `butil.buffersAreEqual` because it checks to see if code is an emailCode. Without this check it would verify the account email even if invalid token found. Not sure the reason why `butil.buffersAreEqual` was used instead of `code === emailCode`.

shane-tomlinson commented on lib/routes/account.js:
```diff
@@ -946,10 +1013,33 @@ module.exports = function (
+            return db.sessionTokenWithVerificationStatus(sessionToken.id)
+              .then(function (token) {
+
+                var sessionVerified = token.tokenVerified
```
I just remembered @philbooth and I talking about this early on - the semantics of this endpoint have changed slightly, which is causing [this test failure in the fxa-js-client](https://github.com/mozilla/fxa-js-client/pull/203#issuecomment-222099775). If the user signs in w/o requesting keys and then calls `/recovery_email/status`, the response contains `verified: false`. This is because sessions that do not request keys are "unverified" by default, so that they cannot be used to get at Sync data.

In real life, returning `verified: false` shouldn't be a problem because `/recovery_email/status` is called after another endpoint indicates a user must verify their email. There is not really much of a reason to call `/recovery_email/status` if the other endpoints indicated the user was verified. Of course, there are some huge assumptions there - I have no idea when Fx Desktop, Fennec, iOS or Fx OS call `/recovery_email/status`.

If there a way to maintain backwards compatibility here, where we return `verified: false` only for accounts that are unverified, or sessions that request keys that have not yet confirmed the signin? I don't know what kind of shenanigans that would require.

vbudhram replied:
Yea I am not sure what kind of shenanigans this would cause either lol. My initial impression is that it wouldn't cause any because it seems the only reason you would have to be calling `/recovery_email/status` is because you want to know that a session/email is verified.

I could see this tweaked a little to better maintain backwards compatibility. If the response looked like

```javascript
{
 verified: "email verified value",
 sessionVerified: "session verified value"
}
```
For future clients, they would have to make sure both are true before proceeding. Not sure if that is worthwhile or makes sense.

shane-tomlinson replied:
```
{
 verified: "email verified value",
 sessionVerified: "session verified value"
}
```

Unfortunately, that won't work. All existing clients look only at the "verified" field to decide whether to show a "go verify your email" screen, which we want to happen if either the email or session need to be verified. The browser also looks at that field to determine whether it needs to display the "Go verify your email" button in about:preferences#sync and the warning sign in the hamburger menu.

shane-tomlinson commented on lib/routes/account.js:
```diff
@@ -972,27 +1062,42 @@ module.exports = function (
+              .then(func.bind(
+                mailer,
+                sessionToken,
+                doSigninConfirmation ? token.tokenVerificationId : sessionToken.emailCode, // Send tokenVerficationId, unless sign-in confirmation is disabled
```
There is a subtle problem here, as noted in the [Sign up using old auth server, with new auth server, click "resend" from "/confirm" screen test notes](https://public.etherpad-mozilla.org/p/fxa-83-confirm-login). If the user signs up with the old auth server, then clicks "resend" from the confirm screen under the new auth server, `sendVerifyCode` is called with a code of `null`. I think this line should be something along the lines of:

```js
doSigninConfirmation ? token.emailVerified ? token.tokenVerificationId: sessionToken.emailCode : sessionToken.emailCode
```

Perhaps that'd be easier to read if the code was set up where `func` is set.

shane-tomlinson replied:
This seems to fix the problem, though I'm not sure of the repercussions:

```diff
diff --git a/lib/routes/account.js b/lib/routes/account.js
index 8ea6ae7..5a9a35d 100644
--- a/lib/routes/account.js
+++ b/lib/routes/account.js
@@ -1068,6 +1072,7 @@ module.exports = function (
 
             // Choose which type of email to send
             var doSigninConfirmation = requestHelper.shouldEnableSigninConfirmation(sessionToken.uid.toString('hex'), config)
+            var code
             var func
             if (!token.emailVerified || !doSigninConfirmation) {
 
@@ -1076,10 +1081,15 @@ module.exports = function (
                 return reply({})
               }
 
+              code = sessionToken.emailCode
               func = mailer.sendVerifyCode
             } else {
+              code = sessionToken.tokenVerificationId
               func = mailer.sendVerifyLoginEmail
             }
 
             return customs.check(
               request.app.clientAddress,
@@ -1088,7 +1098,7 @@ module.exports = function (
               .then(func.bind(
                 mailer,
                 sessionToken,
-                doSigninConfirmation ? token.tokenVerificationId : sessionToken.emailCode, // Send tokenVerficationId, unless sign-in confirmation is disabled
+                code,
                 userAgent.call({
                   service: service,
                   timestamp: Date.now(),
```

vbudhram replied:
I can't think of any repercussions of this either. My original thinking was that in the new auth-server, if we always send a `tokenVerificationId`, the session and email get verified. While it is less code, it doesn't cover the scenario above. I can update this function.

philbooth commented on lib/routes/utils/request_helper.js:
```diff
@@ -9,9 +9,32 @@
+  if (config.signinConfirmation && config.signinConfirmation.enabled) {
+
+    // Returns the first 2 integers of uid
+    var uidNum = parseInt(uid.substr(0, 4), 16) % 100
```
This seems to rely on the most significant digits. Do we know whether, in practice, that will return an even spread of values to base the sample rate on?

By definition, as numbers get bigger, the most significant digits will vary less and less. You get far more entropy from the least significant digits. Since we're already using a modulo operation here, that can be achieve by just using `parseInt(uid) % 100` and ditching the `substr`.

Does that make sense or am I worrying about nothing here?

philbooth replied:
Just read that back to myself and it sounds like rubbish now I think about it. Ignore me, sorry. 😊

rfk replied:
These uids are v4 uuids [1], and if I understand [2] correctly that means their leading bits should be uniformly random.  Definitely worth double-checking though!

[1] https://github.com/mozilla/fxa-auth-server/blob/master/lib/routes/account.js#L152
[2] https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_.28random.29

philbooth replied:
Cool, thanks. To properly satisfy myself, I tried it in the repl. Looks pretty uniform to me!

```
> const distribution = new Array(100).fill(0)
undefined
> const uuid = require('uuid')
undefined
> for (var i = 0; i < 1000000; ++i) {
... ++distribution[parseInt(uuid.v4('binary').toString('hex').substr(0, 4), 16) % 100]
... }
9917
[ 10007,
  10078,
  9856,
  9901,
  10072,
  10006,
  9916,
  9832,
  10006,
  10190,
  10103,
  9941,
  10034,
  9811,
  10026,
  10084,
  9804,
  9911,
  10031,
  9927,
  9860,
  10069,
  9857,
  9722,
  10082,
  9962,
  10054,
  10125,
  10114,
  10081,
  10076,
  10024,
  9946,
  10029,
  10002,
  10193,
  9943,
  10100,
  9992,
  9890,
  10027,
  9883,
  9991,
  10165,
  9929,
  10013,
  9948,
  10138,
  9966,
  9985,
  10146,
  9745,
  10156,
  10020,
  9881,
  9859,
  10017,
  10068,
  10124,
  10045,
  9905,
  10134,
  10038,
  10098,
  10051,
  10073,
  10134,
  10064,
  10086,
  9972,
  9914,
  9830,
  9923,
  9917,
  10003,
  10047,
  9837,
  9950,
  9945,
  10017,
  9859,
  10100,
  9944,
  10171,
  10132,
  10108,
  9946,
  10034,
  9917,
  9999,
  9972,
  9828,
  9992,
  9992,
  10051,
  10126,
  9970,
  10130,
  10073,
  9955 ]
```

vbudhram commented on lib/routes/account.js:
```diff
@@ -1046,31 +1043,15 @@ module.exports = function (
 
         function createResponse() {
 
-          var doSigninConfirmation = requestHelper.shouldEnableSigninConfirmation(sessionToken.uid.toString('hex'), config)
```
@shane-tomlinson @philbooth Per our conversation yesterday, I update `/recovery_email/status` to not reference any sign-in confirmation values. In theory, this endpoint should return the correct value for `verified` regardless if the feature is enabled or not. 

However one thing I noticed while adding this in, since there is no metricsContext, I wouldn't be able to determine if this request is coming from a user that has the sign-in feature enabled and thus return `emailVerified` and `sessionVerified` in the response. The old auth server does not have these additional fields.

cc @rfk

rfk replied:
I think it might be OK to just return these extra fields regardless of whether the user has sign-in confirmation feature enabled.  Old code will ignore them, new code will see not-opted-in-yet users as being verified on both counts.

vbudhram replied:
Sounds good. @shane-tomlinson any repercussions on the content server for this?

philbooth commented on lib/routes/account.js:
```diff
@@ -1095,41 +1077,40 @@ module.exports = function (
 
-        return db.sessionTokenWithVerificationStatus(sessionToken.id)
-          .then(function (token) {
+        var doSigninConfirmation = requestHelper.shouldEnableSigninConfirmation(sessionToken, config, request)
```
Do we need to call `shouldEnableSigninConfirmation` here?

It seems to me that we can just check the existence of `sessionToken.tokenVerificationId`. If this user wasn't part of the cohort, `tokenVerificationId` wouldn't have been specified at the point of creation. Right? Or wrong?

philbooth commented on docs/api.md:
```diff
@@ -786,9 +790,9 @@ Failing requests may be due to the following errors:
 Returns the "verified" status for the account's recovery email address.
 
-Currently, each account is associated with exactly one email address. This address must be "verified" before the account can be used (specifically, `/v1/certificate/sign` and `/v1/account/keys` will return errors until the address is verified). In the future, this may be expanded to include multiple addresses, and/or alternate types of recovery methods (e.g., SMS). A new API will be provided for  ...
+Currently, each account is associated with exactly one email address. This address and the sessionToken must be "verified" before the account can be used (specifically, `/v1/certificate/sign` and `/v1/account/keys` will return errors until they are verified). In the future, this may be expanded to include multiple addresses, and/or alternate types of recovery methods (e.g., SMS). A new API will b ...
```
The bit about `/certificate/sign` returning an error for unverified tokens is wrong, right?

vbudhram replied:
Nice catch! Looks like `/v1/certifcate/sign` uses the `sessionTokenWithDevice` strategy, which doesn't check for verification status, will update.

philbooth commented on lib/routes/account.js:
```diff
@@ -194,14 +197,20 @@ module.exports = function (
+          doSigninConfirmation = requestHelper.shouldEnableSigninConfirmation(account, config, request)
+
+          // Create a verified session if preVerified flag set or sign-in confirmation disabled
+          var tokenVerifiedId = (preVerified || !doSigninConfirmation) ? undefined : tokenVerificationId
```
Using a fresh `var` with a slightly-different name feels a bit funny to me here. Would it be clearer instead if the condition was reworked to simply unset `tokenVerificationId`? Something like:

```js
if (preVerified || ! doSigninConfirmation) {
  tokenVerificationId = undefined
}
```

rfk commented on config/index.js:
```diff
@@ -387,6 +387,39 @@ var conf = convict({
+      default: 1.0,
+      env: 'SIGNIN_CONFIRMATION_RATE'
+    },
+    allowClients: {
```
nit: "supportedClients" and "supportedEmails" may be slightly clearer names for these config values.

rfk replied:
To which I should add, 👍 on pulling these out into nice easy to follow config values!

rfk commented on config/index.js:
```diff
@@ -387,6 +387,39 @@ var conf = convict({
+  signinConfirmation: {
+    enabled: {
+      doc: 'enable signin confirmation',
+      default: true,
```
I wonder if it's worth defaulting this to false, so that we can't accidentally deploy with it turned on full-throttle for the initial version.  Thoughts?

rfk commented on docs/api.md:
```diff
@@ -786,9 +790,9 @@ Failing requests may be due to the following errors:
 Returns the "verified" status for the account's recovery email address.
 
-Currently, each account is associated with exactly one email address. This address must be "verified" before the account can be used (specifically, `/v1/certificate/sign` and `/v1/account/keys` will return errors until the address is verified). In the future, this may be expanded to include multiple addresses, and/or alternate types of recovery methods (e.g., SMS). A new API will be provided for  ...
+Currently, each account is associated with exactly one email address. This address must be "verified" before the account can be used (specifically, `/v1/account/keys` will return errors until verified). In the future, this may be expanded to include multiple addresses, and/or alternate types of recovery methods (e.g., SMS). A new API will be provided for this extra functionality.
```
I'm not sure this needs to change actually, IIUC `/certificate/sign` should still fail if the *email* is not verified, which is that this section is talking about.

vbudhram replied:
Wonder if I am missing something, looking at the `/certificate/sign` endpoint, it doesn't appear to reject based on email status, https://github.com/mozilla/fxa-auth-server/blob/master/lib/routes/sign.js#L95

rfk replied:
It's further up in the file, at https://github.com/mozilla/fxa-auth-server/blob/master/lib/routes/sign.js#L44

rfk commented on lib/db.js:
```diff
@@ -724,6 +750,16 @@ module.exports = function (
+        )
+    )
+  }
+
```
I believe this has now reverted to being "verifyTokens" plural, as of https://github.com/mozilla/fxa-auth-db-mysql/commit/5a45e40e90edcf261be0e9b024e2fedc432e7e91

rfk replied:
In fact, I suggest updating the npm-shrinkwrap file as part of this PR, so that it pulls in an appropriate commit of fxa-auth-db-mysql with the latest changes.

rfk commented on lib/routes/account.js:
```diff
@@ -382,9 +436,7 @@ module.exports = function (
         function sendNewDeviceLoginNotification () {
-          if (config.newLoginNotificationEnabled && requestHelper.wantsKeys(request)) {
-            // The response doesn't have to wait for this,
-            // so we don't return the promise.
```
Please keep the comment about how we purposely don't wait for the promise (unless you want to start waiting for the promise, e.g. to simplify tests)

rfk commented on lib/tokens/session_token.js:
```diff
@@ -22,6 +22,10 @@ module.exports = function (log, inherits, Token) {
+
+    // Tokens are considered verified if no tokenVerificationId exists
+    this.tokenVerificationId = details.tokenVerificationId || null
+    this.tokenVerified = this.tokenVerificationId ? false : true
```
👍

rfk commented on lib/routes/account.js:
```diff
@@ -1004,26 +1078,35 @@ module.exports = function (
+          func = mailer.sendVerifyCode
+        } else {
+          code = sessionToken.tokenVerificationId
+          func = mailer.sendVerifyLoginEmail
```
The branching logic here is not obvious to me.  What is `!sessionToken.emailVerified || sessionToken.tokenVerified` guarding against?  My naive impression is that sending the emailCode is never the right thing to do, but I could easily be missing something.

rfk replied:
FWIW my impression of the logic was to be something like:

```
if (sessionToken.emailVerified && sessionToken.tokenVerified) {
  return reply({})
}

if (sessionToken.tokenVerificationId) {
  code = sessionToken.tokenVerificationId
} else {
  code = sessionToken.emailCode
}

if (sessionToken.emailVerified) {
  func = mailer.sendVerifyCode
} else {
  func = mailer.sendVerifyLoginEmail
}
```

vbudhram replied:
I believe this was my attempt to handle a possible rollback of the content server issuing a request to the new auth-server. In which case it would send the verify code email. Your approach works well too.

rfk commented on lib/routes/account.js:
```diff
@@ -1045,50 +1128,84 @@ module.exports = function (
+                    tokenVerificationId: code
+                  })
+                })
+                .catch(function (err) {
```
Should we check that the `err` here is in fact the "not found" error rather than e.g. an operational database error?

vbudhram replied:
Yea, I think we should. While adding this in, it reminded me of an issue that @shane-tomlinson mentioned a while back that slipped out of my memory. Namely, that fxa-db-mysql returns 116 for [not found](https://github.com/mozilla/fxa-auth-db-mysql/blob/master/fxa-auth-db-server/lib/error.js#L31), but in fxa-auth-server, 116 is [endpoint not supported](https://github.com/mozilla/fxa-auth-server/blob/master/lib/error.js#L17).

Do errno in fxa-mysql and fxa-auth server have to be related or should I add a function to Errors to relate fxa-mysql and fxa-auth errno?

rfk replied:
Yeah, I don't like that the error numbers are different across different repos.  The simplest approach, may be to add code in ./lib/db.js to from auth-db-mysql error numbers back to auth-server error numbers.  There's already a little bit of such logic in that file:

  https://github.com/mozilla/fxa-auth-server/blob/master/lib/db.js#L764

rfk commented on lib/routes/account.js:
```diff
@@ -1045,50 +1128,84 @@ module.exports = function (
+                    uid: account.uid,
+                    locale: account.locale,
+                    tokenVerificationId: code
+                  })
```
Events logged by `log.event`, go out over SQS to listeners such a sync, basket, etc.  Unless we have a specific use-case for one of them consuming this `tokenVerified` event, I don't think we need to log one here.  And either way, we probably don't need to include the tokenVerificationId.

rfk commented on lib/routes/account.js:
```diff
@@ -1045,50 +1128,84 @@ module.exports = function (
-                      )
-                    }
+                  // Completely unknown code
+                  return err
```
Is this dead code?  Both branches in the above `if` statement terminate control flow.

rfk replied:
(even if not, you probably need to throw it rather than return it)

rfk commented on lib/routes/password.js:
```diff
@@ -149,7 +149,7 @@ module.exports = function (
         var account, verifyHash, sessionToken, keyFetchToken
 
-        return changePassword()
+        changePassword()
```
I'm suspicious of losing the `return` here, do we need to?

vbudhram replied:
I don't believe we need it, https://github.com/mozilla/fxa-auth-server/blob/edecb24ec75b6193bf2bda0ecdb2476801a5a427/lib/routes/password.js#L49.

rfk replied:
No worries - my only concerns was if we needed it for e.g. tests, but if it's not breaking anything, I don't really mind :-)

rfk commented on lib/routes/password.js:
```diff
@@ -204,22 +204,21 @@ module.exports = function (
+            email: account.email,
+            emailCode: account.emailCode,
+            emailVerified: account.emailVerified,
+            verifierSetAt: account.verifierSetAt
```
Despite the comment about using the verification status of the current session, this doesn't appear to take any state from the provided sessionToken.  Also, the `sessionToken` argument is still optional, so perhaps we should maintain the existing logic of only creating one if one was provided in the request.

vbudhram replied:
Nice catch! Will update accordingly.

rfk commented on lib/routes/account.js:
```diff
@@ -229,34 +241,39 @@ module.exports = function (
+                    uid: account.uid,
+                    kA: account.kA,
+                    wrapKb: wrapKb,
+                    emailVerified: account.emailVerified
```
There was discussion about making the keyFetchTokens also require verification here.  Just to be clear, we we deferring that to follow-up work because it's not needed for the desktop clients?  If so please file a follow-up issue so that we don't lost track of it.

vbudhram replied:
Follow up issue created, https://github.com/mozilla/fxa-auth-server/issues/1287

rfk commented:
It's not explicit in this PR, but IIUC, sessionTokens created through the /account/reset endpoint will be verified by default, and this is as it should be since the user confirmed their email when they did the password reset.  So I think we need to add:

* An explicit comment to that effect, at the point in the /account/reset route where we create the sessionToken
* An explicit test that this is working as intended

rfk commented on lib/routes/utils/request_helper.js:
```diff
@@ -9,9 +9,52 @@
  */
 function wantsKeys (request) {
-  return request.query.keys === 'true'
+  return request.query.keys === 'true' || request.query.keys === true
```
Is hapi doing some parsing to turn the query param into a boolean `true` value?

vbudhram replied:
This was a self inflicted bug introduced when using the mocked requests :(, fixed.

rfk commented on lib/routes/utils/request_helper.js:
```diff
@@ -9,9 +9,52 @@
+
+    // If feature enabled, always enable for these emails
+    var emailDomain = account.email.substring(email.indexOf('@'), email.length).toLocaleLowerCase()
+    var isValidEmail = config.signinConfirmation.allowEmails.indexOf(emailDomain) > -1
```
Ah, interesting, when I read the name "allowEmails" I thought it meant "only allow it for these emails", not "always enable it for these emails".  We sometimes use the word "force" in this sort of context, e.g. using `forceDeviceList=true` to show the device list UI.  So I wonder if something like "supportedClients" and "forcedEmails" might make it a little clearer what these config values mean.

rfk commented on lib/routes/utils/request_helper.js:
```diff
@@ -9,9 +9,52 @@
+    var isValidEmail = config.signinConfirmation.allowEmails.indexOf(emailDomain) > -1
+    if (isValidEmail) {
+      return true
+    }
```
I think we need to move this check down after the `isValidContext` check, otherwise @mozilla.com users logging in on their Android device are going to have a bad time.

rfk replied:
More broadly, this function might be easier to follow as a single level of short-circuiting returns, rather than nested if-elses.  For example:

```
if (it's not enabled in the config) {
  return false
}

if (the context value is not in the list of supported clients) {
  return false
}

if (the email is on the list of forced emails) {
  return true
}

calculate the sample rate
return sample rate < the configured rate
```

rfk commented on lib/server.js:
```diff
@@ -57,6 +57,22 @@ function create(log, error, config, routes, db) {
+          cb
+        )
+    }
+  }
```
It's not obvious why we need a separate `makeVerifiedCredentialsFn` rather than re-using the almost identical `makeCredentialsFn` - is this a holder from when the strategy for explicitly checking the verification status rather than returning it?

rfk commented on test/client/index.js:
```diff
@@ -113,6 +113,31 @@ Client.createAndVerify = function (origin, email, password, mailbox, options) {
     )
 }
 
+Client.loginAndVerify = function (origin, email, password, mailbox, options) {
```
nice 👍

rfk commented on test/local/account_routes.js:
```diff
@@ -161,6 +165,104 @@ test(
+      })
+    })
+    .then(function(response) {
+      t.equal(mockDB.deleteAccount.callCount, 0)
```
Having a spy to test that `deleteAccount` was never called seems a little odd, do we have reason to worry that it might be called?  If so, adding a comment here somewhere would be helpful.

vbudhram replied:
This can be safely removed, there is an explicit test case above, https://github.com/mozilla/fxa-auth-server/blob/edecb24ec75b6193bf2bda0ecdb2476801a5a427/test/local/account_routes.js#L95

rfk commented on test/local/account_routes.js:
```diff
@@ -161,6 +165,104 @@ test(
+          uid: uuid.v4('binary').toString('hex'),
+          email: TEST_EMAIL,
+          emailVerified: true,
+          tokenVerified: true
```
It would be interesting to add one of these tests with `emailVerified: true` but `tokenVerified: false`, to confirm that it reflects the expected behaviour.

rfk commented on test/local/account_routes.js:
```diff
@@ -619,6 +721,254 @@ test(
+    })
+      .then(function (response) {
+        t.equal(mockMailer.sendNewDeviceLoginNotification.callCount, 1, 'mailer.sendNewDeviceLoginNotification was called')
+        t.equal(mockMailer.sendVerifyLoginEmail.callCount, 0, 'mailer.sendVerifyLoginEmail was called')
```
was *not* called

rfk commented on test/local/push_db_tests.js:
```diff
@@ -101,6 +102,7 @@ test(
         })
         .then(function(emailRecord) {
           emailRecord.createdAt = Date.now()
+          emailRecord.tokenVerificationId = ACCOUNT.tokenVerificationId
```
It's not obvious why we need to add zeroed-out tokenVerificationIds here - can't we create the token with no id and have things continue to work?

rfk commented on test/local/verifier_upgrade_tests.js:
```diff
@@ -75,7 +75,7 @@ createDBServer().then(
         function (server) {
           var client
-          return Client.login(config.publicUrl, email, password, server.mailbox, {keys:true})
+          return Client.login(config.publicUrl, email, password, server.mailbox)
```
Why did we need to move the location of `keys: true` in this test?

vbudhram replied:
I think this was an artifact of having sign-in confirmation enabled by default, which only gets triggered when requesting keys.

rfk commented on test/mocks.js:
```diff
@@ -64,5 +135,8 @@ function mockObject (methodNames) {
   spyLog: spyLog,
+  createDB: createDB,
+  createMailer: createMailer,
+  createRequest: createRequest,
```
Nice refactor and cleanup on this mocking code!  Is it worth using the new `mockObject` helper here to make them even more flexible?

Naming nit: for consistency with the existing exports, should these be something `mockDB`, `mockMailer`, etc?

rfk commented on test/remote/account_create_tests.js:
```diff
@@ -35,19 +34,24 @@ TestServer.start(config)
+      return Client.create(config.publicUrl, email, password)
         .then(
-          function (status) {
-            t.equal(status.verified, false)
```
What happened to this check on `status.verified`?

vbudhram replied:
I can add it back in, just a case where it was being tested in other areas and didn't want to re-test it.

rfk replied:
I don't mind leaving it out, just as long as we're cross-checking that we don't lose coverage on anything

rfk commented on test/remote/account_preverified_token_tests.js:
```diff
@@ -4,11 +4,11 @@
 
 process.env.TRUSTED_JKUS = 'http://127.0.0.1:9000/.well-known/public-keys'
-var config = require('../../config').getProperties()
+var config = require('../../config/index').getProperties()
```
What happened to require us to refer explicitly to `index` here?

vbudhram replied:
Not 100% sure, but think it was added in because the tests were failing originally with sign-in confirmation. Retested and it works without it, removing.

rfk commented on test/remote/account_preverified_token_tests.js:
```diff
@@ -102,13 +102,6 @@ TestServer.start(config)
-        .then(
-          function (keys) {
-            t.ok(Buffer.isBuffer(keys.kA), 'kA exists')
-            t.ok(Buffer.isBuffer(keys.wrapKb), 'wrapKb exists')
```
I think we should keep the explicit tests for retreiving keys here, since signingup against an unverified account has magic behaviour including re-creating a fresh set of keys.

rfk commented on test/remote/account_signin_verification_tests.js:
```diff
@@ -0,0 +1,258 @@
+  .then(function main(server) {
+
+    test(
+      'account signin with without keys does not set challenge',
```
"with without"

rfk commented on test/remote/account_signin_verification_tests.js:
```diff
@@ -0,0 +1,258 @@
+            }
+          )
+      }
+    )
```
Great, I'm really pleased to see an explicit test for this case.

rfk commented on test/remote/certificate_sign_tests.js:
```diff
@@ -42,6 +42,7 @@ TestServer.start(config)
             t.ok(payload['fxa-generation'] > 0, 'cert has non-zero generation number')
             t.ok(new Date() - new Date(payload['fxa-lastAuthAt'] * 1000) < 1000 * 60 * 60, 'lastAuthAt is plausible')
             t.equal(payload['fxa-verifiedEmail'], email, 'verifiedEmail is correct')
+            t.equal(payload['fxa-tokenVerified'], true, 'tokenVerified is correct')
```
It's critical that non-verified sessions produce `fxa-tokenVerified === false` in the certificate, so I think it would be worth adding an explicit test case that to exercise it, i.e. one that signs a certificate using an unverified session token.

vbudhram replied:
Added here https://github.com/mozilla/fxa-auth-server/blob/3505d6db5f7e48f4eb2598e3e3d507c2bccb2aa1/test/remote/account_signin_verification_tests.js#L400

rfk commented on test/remote/recovery_email_resend_code_tests.js:
```diff
@@ -0,0 +1,88 @@
+            }
+          )
+      }
+    )
```
An interesting additional testcase for this file, might be to generate two client objects, have them both login and resend the email, and assert that we get a different code in the email (because they're using two different session tokens)

rfk replied:
Oh, looks like you have something very similar in recovery_email_verify_tests.js already

rfk commented:
@vbudhram I feel like I left a *lot* of comment just now, but this is very very close to landing.  Nice work!

vbudhram commented:
Hey @rfk, sorry for delay but I have finished updating this PR from your review. Here is a quick summary of a few changes

- [x] Update config values, uses `forceEmails` and `supportedClient`
- [x] Sign-in confirmation disabled by default
- [x] Updated api docs
- [x] Cleanup branching logic for `/recovery_email/resend_code`
- [x] Explict check for not found in `/recovery_email/verify_code`
- [x] Create sessionToken with correct verification status in `/password/change/finish`
- [x] Reorganized `shouldEnableSigninConfirmation` logic
- [x] Removed makeVerifiedCredentialsFn function
- [x] Added extra unit test for `/cert/sign`

Would you mind doing another look over? Thanks!

rfk commented on lib/routes/account.js:
```diff
@@ -1314,8 +1428,8 @@ module.exports = function (
-            // Create a sessionToken so that the client does
-            // not have to re-login
+            // Since the only way to reach this point is clicking a
+            // link from the user's email, we create a verified sessionToken
```
👍

rfk commented on lib/routes/password.js:
```diff
@@ -147,15 +150,47 @@ module.exports = function (
 
+        function getSessionVerificationStatus() {
+          if (sessionTokenId) {
+            return Tokens.SessionToken.fromHex(sessionTokenId)
```
I didn't realize the client was passing in the raw sessionToken data blob here, rather than just the token id.  We generally try to avoid passing the raw sessionToken data over the wire because it's a plaintext credential....sorry, I should have noticed this in review of the original change-password PR.  I think we can push ahead with it for now but will file a follow-up issue to address this.

rfk commented:
I suggested a couple of nits as a PR-to-your-PR in https://github.com/mozilla/fxa-auth-server/pull/1288

Otherwise, r+, let's land this thing! 🎆

shane-tomlinson commented:
Seeing this merged brought a tear of joy to my eye.
